orderbook/src/database/app_data.rs

```diff
@@ -0,0 +1,238 @@
+use super::*;
+use anyhow::Result;
+use futures::stream::TryStreamExt;
+use futures::StreamExt;
+use model::app_data::AppDataBlob;
+use primitive_types::{H160, H256};
+use std::borrow::Cow;
+
+#[async_trait::async_trait]
+pub trait AppDataStoring: Send + Sync {
+    async fn insert_app_data(&self, app_data: &AppDataBlob) -> Result<H256, InsertionError>;
+    fn app_data<'a>(&'a self, filter: &'a AppDataFilter) -> BoxStream<'a, Result<AppDataBlob>>;
+}
+
+/// Any default value means that this field is unfiltered.
+#[derive(Clone, Default)]
+pub struct AppDataFilter {
+    pub app_data_hash: Option<H256>,
+    pub referrer: Option<H160>,
+}
+
+#[derive(Debug)]
+pub enum InsertionError {
+    ParsingStringError(serde_json::Error),
+    DuplicatedRecord(H256),
+    AnyhowError(anyhow::Error),
+    DbError(sqlx::Error),
+}
+
+impl From<sqlx::Error> for InsertionError {
+    fn from(err: sqlx::Error) -> Self {
+        Self::DbError(err)
+    }
+}
+
+impl From<serde_json::Error> for InsertionError {
+    fn from(err: serde_json::Error) -> Self {
+        Self::ParsingStringError(err)
+    }
+}
+
+impl From<anyhow::Error> for InsertionError {
+    fn from(err: anyhow::Error) -> Self {
+        Self::AnyhowError(err)
+    }
+}
```

Comment:
you can use `thiserror` crate to derive some of this for you. Alternatively, you can just return an `anyhow::Error` unless you need to match on the specific error in the caller.